src/org/jitsi/impl/neomedia/transform/fec/FlexFec03Receiver.java

```diff
@@ -0,0 +1,341 @@
+                                 RawPacket dest, int destOffset,
+                                 int payloadLength)
+        {
+            for (int i = 0; i < payloadLength; ++i)
+            {
+                dest.getBuffer()[destOffset + i] ^=
+                    source.getBuffer()[sourceOffset + i];
+            }
+        }
+
+        /**
+         * Do the final work when recovering an RTP packet (set the RTP version,
+         * the length, the sequence number, and the ssrc)
+         * @param fecPacket the fec packet
+         * @param recoveredPacket the media packet which was recovered
+         */
+        private boolean finishPacketRecovery(FlexFec03Packet fecPacket, RawPacket recoveredPacket)
+        {
+            // Set the RTP version to 2.
+            recoveredPacket.getBuffer()[0] |= 0x80; // Set the 1st bit
+            recoveredPacket.getBuffer()[0] &= 0xbf; // Clear the second bit
+
+            // Recover the packet length, from temporary location.
+            int length = RTPUtils.readUint16AsInt(recoveredPacket.getBuffer(), 2);
+            // The length field used in the xor does not include the header
+            // length, but we want to include the fixed header length when
+            // setting the length on the packet object
+            int lengthWithFixedHeader = length + RawPacket.FIXED_HEADER_SIZE;
+            if (lengthWithFixedHeader > recoveredPacket.getBuffer().length)
+            {
+                logger.error("Length field of recovered packet is larger" +
+                    " than its buffer");
+                return false;
+            }
+
+            recoveredPacket.setLength(lengthWithFixedHeader);
+            recoveredPacket.setSequenceNumber(missingSequenceNumber);
+            recoveredPacket.setSSRC((int)fecPacket.getProtectedSsrc());
+
+            return true;
+        }
+
+        /**
+         * Recover a media packet based on the given FlexFEC packet and
+         * received media packets
+         * @return the recovered packet if successful, null otherwise
+         */
+        private RawPacket recover()
+        {
+            if (!canRecover())
+            {
+                return null;
+            }
+
+            //TODO: use a pool?
+            byte[] buf = new byte[1500];
+            RawPacket recoveredPacket = new RawPacket(buf, 0, 1500);
+            if (!startPacketRecovery(this.fecPacket, recoveredPacket))
+            {
+                return null;
+            }
+            for (Integer protectedSeqNum : fecPacket.getProtectedSequenceNumbers())
+            {
+                if (protectedSeqNum != missingSequenceNumber)
+                {
+                    RawPacket mediaPacket = mediaPackets.get(protectedSeqNum);
+                    xorHeaders(mediaPacket, recoveredPacket);
+                    xorPayloads(
+                        mediaPacket,
+                        RawPacket.FIXED_HEADER_SIZE,
```

Comment:
mediaPacket.getOffset() + RawPacket.FIXED_HEADER_SIZE

A thought: do you think it's worth artificially adding a non-zero offset to the packets in the tests, to make sure that we didn't miss a place?

Reply:
fixed...and yeah i'll take a look at doing that.

Reply:
so it turns out a lot of these don't matter, as we copy the media packet into a new RawPacket, so we get an offset of 0...but i do still think it was a good idea, and i've added some tests to verify it now.